lite2/client.go

```diff
@@ -757,6 +782,65 @@ func (c *Client) fetchHeaderAndValsAtHeight(height int64) (*types.SignedHeader,
 	return h, vals, nil
 }
 
+// fetchMissingTrustedHeader finds the closest height after the
+// requested height and does backwards verification.
+func (c *Client) fetchMissingTrustedHeader(height int64, now time.Time) (*types.SignedHeader, error) {
+	c.logger.Info("Fetching missing header", "height", height)
+
+	trustedHeader, err := c.trustedStore.SignedHeaderAfter(height)
+	if err != nil {
+		return nil, errors.Wrapf(err, "can't get signed header after %d", height)
+	}
+
+	var untrustedHeader *types.SignedHeader
+	for i := trustedHeader.Height - 1; i >= height; i-- {
+		untrustedHeader, err = c.signedHeaderFromPrimary(i)
+		if err != nil {
+			return nil, errors.Wrapf(err, "failed to obtain the header #%d", i)
+		}
+
+		if err := untrustedHeader.ValidateBasic(c.chainID); err != nil {
+			return nil, errors.Wrap(err, "untrustedHeader.ValidateBasic failed")
+		}
+
+		if untrustedHeader.Height >= trustedHeader.Height {
+			return nil, errors.Errorf("expected older header height %d to be less than one of newer header %d",
+				untrustedHeader.Height,
+				trustedHeader.Height)
+		}
```

Comment:
`if untrustedHeader.Height >= trustedHeader.Height` - would this ever happen that the primary returns a different height then the one you requested for. In this case shouldn't the `signedHeaderFromPrimary` function filter for that.

Reply:
Good idea. Let's add this check